### PR TITLE
Add function inlining in a later optimizer pass

### LINKED ISSUE #22085: Fix bug when KEY_BASED_SAMPLING_ENABLED=true and inline_sql_functions=true
[key_sampling_percent](https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/operator/scalar/sql/SimpleSamplingPercent.java) is a SQL function.
When we set `inline_sql_functions=true`, we expect the coordinator would always inline this function and we would not need an implementation in the worker eval (Prestissimo in this case).
However, KeyBasedSampler runs after sql functions have already been inlined, so the function doesn't get inlined. 

## Possible Solution
Fix is to add another pass of inline sql functions at the end.

## Steps to Reproduce

presto:tpch> set session inline_sql_functions=true;
presto:tpch> set session KEY_BASED_SAMPLING_ENABLED=true;
presto:tpch> select count(1) from orders join lineitem using(orderkey);

Query 20240304_044118_00007_4g98f failed:  Scalar function name not registered: presto.default.key_sampling_percent, called with arguments: (VARCHAR).

### PR DESCRIPTION
**Pull request by ZacBlanco**
## Description

Adds a function inlining pass to the optimizer after the KeyBasedSampling rule in order to make sure that any rewritten functions are inlined.

## Motivation and Context

Fixes #22085

## Impact

N/A

## Test Plan

Added a test case

## Contributor checklist

- [x] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [x] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [x] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [x] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [x] Adequate tests were added if applicable.
- [x] CI passed.

## Release Notes

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 1]
```diff
@@ -634,7 +634,15 @@ public PlanOptimizers(
...
                         estimatedExchangesCostCalculator,
-                        ImmutableSet.of(new RemoveRedundantIdentityProjections())));
+                        ImmutableSet.of(new RemoveRedundantIdentityProjections())),
+                new IterativeOptimizer(
+                        metadata,
+                        ruleStats,
+                        statsCalculator,
+                        estimatedExchangesCostCalculator,
+                        ImmutableSet.<Rule<?>>builder()
+                                .addAll(new InlineSqlFunctions(metadata, sqlParser).rules())
```
Comment (ClarenceThreepwood) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:644]: Do we still need the earlier pass?

Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:644]: I wasn't sure at first but I briefly skimmed all the rules between the original and new location and I feel pretty confident it can be moved rather than copied

Comment (tdcmeehan) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:636]: We might need to keep both--if we don't inline SQL functions at the beginning, then we might lose downstream opportunities to optimize them.

Comment (aditi-pandit) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:636]: Agree that keeping the earlier pass is better.

**Comment by majetideepak** [comment]
@ZacBlanco can you confirm the test you added reproduces the bug without your fix?

**Comment by ZacBlanco** [comment]
I can confirm that the test does not pass without this change

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java [hunk 2]
```diff
@@ -855,13 +863,13 @@ public PlanOptimizers(
         // Optimizers above this do not need to care about aggregations with the type other than SINGLE
         // This optimizer must be run after all exchange-related optimizers
         builder.add(new IterativeOptimizer(
-                metadata,
```
Comment (tdcmeehan) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java:858]: Unrelated change?

**Review by tdcmeehan — APPROVED** [review]

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java (hunk 2 above)
Comment (ZacBlanco) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/PlanOptimizers.java]: IntelliJ formatter didn't like this indentation for some reason. Fixed

**Review by tdcmeehan — DISMISSED** [review]

**Review by majetideepak — DISMISSED** [review]
Thanks, @ZacBlanco

**Review by tdcmeehan — APPROVED** [review]

**Review by rschlussel — APPROVED** [review]
